OEPS/OEP-5.mediawiki

```diff
@@ -57,6 +57,18 @@ Returns the number of NFTs assigned to <code>address</code>.
 The parameter <code>address</code> must be a 20-byte address.
 If not, this method must <code>throw</code> an exception.
 
+====tokensOf====
+
+<pre>
+def tokensOf(address)
+</pre>
+
+Returns a list of the tokenIds of the NFTs assigned to <code>address</code>.
```

Comment:
neovm has max size limitation for list, which is very annoying.

Reply:
@laizy what is the current size limitation? This standard should enforce pagination then. `tokensOf` should have an optional parameter `page` and should return a list and then something like the total number of pages, or the id of the next page.

Reply:
Oh I see, it's not just the array being returned - but the array being stored in storage. What is the maximum? We could add something that just stores it in numerous arrays.

Reply:
Why is `MAX_ARRAY_SIZE` set to 1024 (https://github.com/ontio/ontology/blob/master/vm/neovm/params.go#L26)? 

Can we change that? What's the reasoning for limiting to 1024? That'll hurt many developers. 4294967295 (UInt32 max value) isn't necessarily required, but I'd imagine at least 100,000,000 items could be required for some people.

Reply:
For reference, in a test on SmartX I was able to build an array with 3500 Integers (in Debug-Run)

Reply:
Unclear where it's being used https://github.com/ontio/ontology/search?q=MAX_ARRAY_SIZE&unscoped_q=MAX_ARRAY_SIZE

Reply:
> Why is `MAX_ARRAY_SIZE` set to 1024 

neovm's opcode is not primitive enough, some opcode execution is `O(n)`, so the size must have some  upper bound to avoid attack.

> Is it 1024 bytes or items in the array?

items
> For reference, in a test on SmartX I was able to build an array with 3500 Integers (in Debug-Run)

I think this is a bug, missing check in [opAppend](https://github.com/ontio/ontology/blob/902dd123d18f82c588bc6c501b959655fa683441/vm/neovm/func_validate.go#L571)

Reply:
> neovm's opcode is not primitive enough, some opcode execution is O(n), so the size must have some upper bound to avoid attack.

What kind of workaround do you think we could implement? Maybe using a linked list kind of structure?

> items

What is the maximum upperbound that wouldn't impact performance? Is it 1024?

> I think this is a bug, missing check in opAppend

Ah gotcha. I opened this issue https://github.com/ontio/ontology/issues/930

Reply:
> What kind of workaround do you think we could implement? Maybe using a linked list kind of structure?

yes, or you can split large list into small chunks.  

> What is the maximum upperbound that wouldn't impact performance? Is it 1024?

1024 is just choose to fit most contract requirement, we can increase this value if it will not lead to hard fork, but it is not a long-term solution. By the way, we are also exploring the `wasm` contract, which will overcome some weakness of `neovm`, but is not ready yet.

Reply:
> yes, or you can split large list into small chunks.

I'm going to look into implementing a linked list and see how that works. I think the small chunks could be more difficult to maintain with the lists changing frequently.

> 1024 is just choose to fit most contract requirement, we can increase this value if it will not lead to hard fork, but it is not a long-term solution. By the way, we are also exploring the wasm contract, which will overcome some weakness of neovm, but is not ready yet.

Very interested to see how this turns out. I agree that changing the 1024 value isn't a solution. I think creating a more efficient method for storage (i.e. a linked list or broken up array) is a better solution.

I'm going to try to implement a doubly linked list and maybe add it directly into the python compiler. Once a solution is reached we should re-evaluate this PR.

Reply:
Doubly linked list doesn't appear to be possible. Having two nodes reference each other causes the vm stack size to be exceeded

Reply:
@laizy instead, could we store the tokenIds in a map? And `tokensOf` would just call `tokens.keys()`, where `tokens` is the dictionary of an address' tokenIds?

What would the best way to delete an item from a dictionary? Is there any limits on `keys()`?

Reply:
> What would the best way to delete an item from a dictionary
`dct.remove(<KEY>)` works

@laizy Is there a limit on how many keys can be stored in a dictionary?

Reply:
`keys` return an array, which will lead to the same problem.

how about this scheme, all data saved in one array, when its size too large, packed it:
```
struct PackedList {
    array []Item   //data layout: [[[,,,,,,],,,,,,],,,,]
    numItems int
}

func (self *PackedList)Push(it Item) {
    if len(self.array) == 1024 {
        temp := self.array
        self.array = NewArray()
        self.array.Push(temp)
        self.array.Push(it)
    }
    self.numItems += 1
}
```

Reply:
If that would work, we need to wait until 1.7.0 is deployed to the mainnet. @laizy when do you think that'll happen?

Reply:
@laizy how would we remove an item from that scheme?

Reply:
> `keys` return an array, which will lead to the same problem.

What would happen if you call `keys` on a dictionary with more than 1024 items?

Reply:
@laizy do the issues with array length only apply to append/remove? What's the issue with `keys()` returning more than 1024 items? Could the solution be removing that limitation?

Reply:
> @laizy how would we remove an item from that scheme?

well, `swap(id, numItems-1)`, and remove the last item. since the item order is not need to remain unchanged.

Reply:
> well, swap(id, numItems-1), and remove the last item. since the item order is not need to remain unchanged.

I'm not sure I understand. If we've added 1025 items to the PackedList, how would we remove the 100th item? Since it'd be `[[ 1 ... 1024 ], 1025 ]`? What would it look like after the remove? Are you assuming the id's are indexes?

Reply:
remove 100th item from `[[ 1 ... 1024 ], 1025 ]`  is :  `[ 1 ..99, 1025, 101, ... 1024 ]`

Reply:
@laizy so I was able to get it to work: https://github.com/WyattMufson/OEP5-PackedList/blob/master/packed.py

We can add and remove items from a packed list like this. This should be enough to solve the constraints for `tokensOf`

Reply:
@laizy will we be limited now by the amount of steps the vm can take?

Reply:
Also I'm sure my PackedList code could use some optimizations

Reply:
Also to clarify, the example code is using a maximum of 2 items in an array (instead of 1024). It's easier to visualize and write out on paper. Changing `MAXIMUM_ARRAY_LENGTH` to 1024 still works.

Reply:
Moved the PackedList example here https://github.com/TowerBuilders/PackedList